<a href="https://colab.research.google.com/github/JuCaRiCo/202602-AnaliticaDescriptivaPredictiva/blob/main/Sesion_Profiling_Estudiante_Gizeh_JCR.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sesión — Tu Propio Perfilador + Profiling Automatizado

**Notebook de práctica — Sesión síncrona (Teams)**

**Agrega los siguientes datos**

- Número de Equipo: N/A
- Nombre de los Integrantes del Equipo:<br/>
 1.Gizeh Anaid Gutierrez Lopez <br/>
 2.Juan Carlos Rivera Cordova<br/>


## Repaso rápido

En Sesión 10 y Sesión 11 hiciste *profiling manual*: revisaste estructura, tipos, duplicados, valores faltantes y variables categóricas de un dataset, celda por celda, con código específico para cada pregunta.

Hoy vas a **empaquetar ese conocimiento en funciones reutilizables** — tu propio mini-perfilador, escrito por ti y aplicado al dataset que le toque a tu equipo.

No es sintaxis nueva — son operaciones que ya usaste sueltas en sesiones anteriores. Aquí las organizas como funciones.

## Preparación

Tu equipo va a trabajar con un dataset distinto al de tus compañeros — el que te asigne tu profesor. Carga tu dataset en la siguiente celda (reemplaza la ruta/URL de ejemplo).


In [1]:
import pandas as pd
import numpy as np

# Reemplaza esta línea con la carga del dataset que te asignó tu equipo
url = 'https://raw.githubusercontent.com/CodingwithO/VideoGameSales/master/vgsales.csv'
df = pd.read_csv(url)
df.shape

(16598, 11)

---
## Parte 1 — Funciones a nivel Dataset

Cada función recibe el DataFrame completo (`df`) y regresa **un solo valor**. Completa el cuerpo de cada función donde dice `# TODO`.

In [8]:
def n_columnas(df):
    """Regresa el número de variables (columnas) del DataFrame."""
    return df.shape[1]


def n_filas(df):
    """Regresa el número de observaciones (filas) del DataFrame."""
    return df.shape[0]


In [3]:
def n_datos_ausentes(df):
    """Regresa el número total de celdas con datos ausentes (NaN) en todo el DataFrame."""
    return df.isna().sum().sum()

def pct_celdas_ausentes(df):
    """Regresa el porcentaje de celdas con datos ausentes sobre el total de celdas
    (filas x columnas) del DataFrame."""
    totalCeldas = len(df) * len(df.columns)

    return (n_datos_ausentes(df) / totalCeldas) * 100

In [4]:
def n_filas_duplicadas(df):
    """Regresa el número de filas duplicadas en el DataFrame."""
    return df.duplicated().sum()

def pct_filas_duplicadas(df):
    """Regresa el porcentaje de filas duplicadas sobre el total de filas."""
    return (n_filas_duplicadas(df) / len(df)) * 100

In [5]:
def n_variables_numericas(df):
    """Regresa el número de columnas con tipo de dato numérico (int o float)."""
    return len(df.select_dtypes(include=['number']).columns)

def n_variables_booleanas(df):
    """Regresa el número de columnas con tipo de dato booleano."""
    return len(df.select_dtypes(include=['bool']).columns)

def n_variables_categoricas(df):
    """Regresa el número de columnas con tipo de dato object o category."""
    return len(df.select_dtypes(include=['object', 'category']).columns)

### Verifica tus funciones

Compara lo que regresan tus funciones contra `.info()` y `.describe()` — deben coincidir.

In [11]:
print('Columnas:', n_columnas(df))
print('Filas:', n_filas(df))
print('Celdas ausentes:', n_datos_ausentes(df))
print(f'% celdas ausentes: {pct_celdas_ausentes(df):.2f}%')
print('Filas duplicadas:', n_filas_duplicadas(df))
print(f'% filas duplicadas: {pct_filas_duplicadas(df):.2f}%')
print('Variables numéricas:', n_variables_numericas(df))
print('Variables booleanas:', n_variables_booleanas(df))
print('Variables categóricas:', n_variables_categoricas(df))

print()
df.info()
print()
df.describe()

Columnas: 11
Filas: 16598
Celdas ausentes: 329
% celdas ausentes: 0.18%
Filas duplicadas: 0
% filas duplicadas: 0.00%
Variables numéricas: 7
Variables booleanas: 0
Variables categóricas: 4

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB



,Rank,Year,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales
count,16598.000000,16327.000000,16598.000000,16598.000000,16598.000000,16598.000000,16598.000000
mean,8300.605254,2006.406443,0.264667,0.146652,0.077782,0.048063,0.537441
std,4791.853933,5.828981,0.816683,0.505351,0.309291,0.188588,1.555028
min,1.000000,1980.000000,0.000000,0.000000,0.000000,0.000000,0.010000
25%,4151.250000,2003.000000,0.000000,0.000000,0.000000,0.000000,0.060000
50%,8300.500000,2007.000000,0.080000,0.020000,0.000000,0.010000,0.170000
75%,12449.750000,2010.000000,0.240000,0.110000,0.040000,0.040000,0.470000
max,16600.000000,2020.000000,41.490000,29.020000,10.220000,10.570000,82.740000


### Función resumen

Combina todas las funciones anteriores en un solo diccionario — tu propio "reporte ejecutivo" del dataset.

In [12]:
def resumen_dataset(df):
    """Regresa un diccionario con el resumen general del DataFrame,
    combinando todas las funciones de la Parte 1."""
    return {
        'n_columnas': n_columnas(df),
        'n_filas': n_filas(df),
        'n_datos_ausentes': n_datos_ausentes(df),
        'pct_celdas_ausentes': pct_celdas_ausentes(df),
        'n_filas_duplicadas': n_filas_duplicadas(df),
        'pct_filas_duplicadas': pct_filas_duplicadas(df),
        'n_variables_numericas': n_variables_numericas(df),
        'n_variables_booleanas': n_variables_booleanas(df),
        'n_variables_categoricas': n_variables_categoricas(df),
    }

resumen_dataset(df)

{'n_columnas': 11,
 'n_filas': 16598,
 'n_datos_ausentes': np.int64(329),
 'pct_celdas_ausentes': np.float64(0.18019695691704368),
 'n_filas_duplicadas': np.int64(0),
 'pct_filas_duplicadas': np.float64(0.0),
 'n_variables_numericas': 7,
 'n_variables_booleanas': 0,
 'n_variables_categoricas': 4}

---
## Parte 2 — Funciones a nivel Columna

Cada función recibe el DataFrame y el **nombre de una columna** (`col`), y regresa un solo valor para esa columna.

In [13]:
def total_ausentes(df, col):
    """Regresa el número de valores ausentes (NaN) en la columna col."""
    return df[col].isna().sum()

def pct_ausentes(df, col):
    """Regresa el porcentaje de valores ausentes en la columna col."""
    return (total_ausentes(df, col) / len(df)) * 100

In [14]:
def media(df, col):
    """Regresa la media de la columna col (asume que es numérica)."""
    return df[col].mean()

def minimo(df, col):
    """Regresa el valor mínimo de la columna col."""
    return df[col].min()

def maximo(df, col):
    """Regresa el valor máximo de la columna col."""
    return df[col].max()

In [15]:
def total_ceros(df, col):
    """Regresa el número de valores iguales a 0 en la columna col."""
    return (df[col] == 0).sum()

def pct_ceros(df, col):
    """Regresa el porcentaje de valores iguales a 0 en la columna col."""
    return (total_ceros(df, col) / len(df)) * 100

In [17]:
def total_negativos(df, col):
    """Regresa el número de valores negativos en la columna col."""
    return (df[col] < 0).sum()

def pct_negativos(df, col):
    """Regresa el porcentaje de valores negativos en la columna col."""
    return (total_negativos(df, col) / len(df)) * 100

### Verifica tus funciones

Elige una columna numérica de tu dataset (idealmente una que tenga valores ausentes o ceros reales) y prueba tus funciones sobre ella.

In [19]:
columna_prueba = 'NA_Sales'  # cambia esto por una columna de tu propio dataset

print('Ausentes:', total_ausentes(df, columna_prueba), f'({pct_ausentes(df, columna_prueba):.2f}%)')
print('Media:', media(df, columna_prueba))
print('Mínimo:', minimo(df, columna_prueba), '| Máximo:', maximo(df, columna_prueba))
print('Ceros:', total_ceros(df, columna_prueba), f'({pct_ceros(df, columna_prueba):.2f}%)')
print('Negativos:', total_negativos(df, columna_prueba), f'({pct_negativos(df, columna_prueba):.2f}%)')

Ausentes: 0 (0.00%)
Media: 0.26466742981082064
Mínimo: 0.0 | Máximo: 41.49
Ceros: 4499 (27.11%)
Negativos: 0 (0.00%)


**Pregunta para discutir en equipo:** de las columnas numéricas de tu dataset, ¿alguna tiene valores negativos o en cero que te parezcan sospechosos? ¿Qué nos dice eso sobre cómo se recolectó ese dato?

### Función resumen

In [25]:
def resumen_columna(df, col):
    """Regresa un diccionario con el resumen de la columna col,
    combinando todas las funciones de la Parte 2."""
    return {
        'total_ausentes': total_ausentes(df, col),
        'pct_ausentes': pct_ausentes(df, col),
        'media': media(df, col),
        'minimo': minimo(df, col),
        'maximo': maximo(df, col),
        'total_ceros': total_ceros(df, col),
        'pct_ceros': pct_ceros(df, col),
        'total_negativos': total_negativos(df, col),
        'pct_negativos': pct_negativos(df, col),
    }

print('Rank:', resumen_columna(df, 'Rank'))
print('Year:',resumen_columna(df, 'Year'))
print('NA_Sales:', resumen_columna(df, 'NA_Sales'))
print('EU_Sales:',resumen_columna(df, 'EU_Sales'))
print('JP_Sales:',resumen_columna(df, 'JP_Sales'))
print('Other_Sales:',resumen_columna(df, 'Other_Sales'))
print('Global_Sales:', resumen_columna(df, 'Global_Sales'))

Rank: {'total_ausentes': np.int64(0), 'pct_ausentes': np.float64(0.0), 'media': np.float64(8300.605253645017), 'minimo': 1, 'maximo': 16600, 'total_ceros': np.int64(0), 'pct_ceros': np.float64(0.0), 'total_negativos': np.int64(0), 'pct_negativos': np.float64(0.0)}
Year: {'total_ausentes': np.int64(271), 'pct_ausentes': np.float64(1.6327268345583803), 'media': np.float64(2006.4064433147546), 'minimo': 1980.0, 'maximo': 2020.0, 'total_ceros': np.int64(0), 'pct_ceros': np.float64(0.0), 'total_negativos': np.int64(0), 'pct_negativos': np.float64(0.0)}
NA_Sales: {'total_ausentes': np.int64(0), 'pct_ausentes': np.float64(0.0), 'media': np.float64(0.26466742981082064), 'minimo': 0.0, 'maximo': 41.49, 'total_ceros': np.int64(4499), 'pct_ceros': np.float64(27.105675382576216), 'total_negativos': np.int64(0), 'pct_negativos': np.float64(0.0)}
EU_Sales: {'total_ausentes': np.int64(0), 'pct_ausentes': np.float64(0.0), 'media': np.float64(0.14665200626581515), 'minimo': 0.0, 'maximo': 29.02, 'total

---
## Cierre — Reporte de alertas de tu dataset

Recorre **todas las columnas** de tu dataset (no solo una) combinando las funciones que construiste con `.nunique()` — que ya conoces de Sesión 10-11 — y completa el reporte de alertas.

In [21]:
# Ayuda: recorre todas las columnas para tener un panorama rápido antes de concluir
for col in df.columns:
    n_unicos = df[col].nunique()
    n_nulos = df[col].isna().sum()
    print(f'{col:25s} -> valores únicos: {n_unicos:6d}  | nulos: {n_nulos:6d}')

Rank                      -> valores únicos:  16598  | nulos:      0
Name                      -> valores únicos:  11493  | nulos:      0
Platform                  -> valores únicos:     31  | nulos:      0
Year                      -> valores únicos:     39  | nulos:    271
Genre                     -> valores únicos:     12  | nulos:      0
Publisher                 -> valores únicos:    578  | nulos:     58
NA_Sales                  -> valores únicos:    409  | nulos:      0
EU_Sales                  -> valores únicos:    305  | nulos:      0
JP_Sales                  -> valores únicos:    244  | nulos:      0
Other_Sales               -> valores únicos:    157  | nulos:      0
Global_Sales              -> valores únicos:    623  | nulos:      0


### Completa el reporte con lo que detectaste en tu dataset

**Atributo(s) con valores en cero:**
_(usa `total_ceros()` / `pct_ceros()` en tus columnas numéricas)_

-JP_Sales: 10,455 ceros/ 62.14%

-Other_Sales: 6,477 ceros / 39.02%

-EU_Sales: 5,730 ceros / 34.52%

-NA_Sales: 4,499 ceros / 27.10%


**Atributo(s) con un solo valor (constante):**
_(busca dónde `.nunique()` es igual a 1)_

— Ninguno

**Atributo(s) con alta cardinalidad:**
_(busca dónde el número de valores únicos se acerca al número de filas — usa la razón valores únicos / total de filas)_

- Rank: 16,598/16,598, esto es debido a que es usado como indice
- Name: 11,493/16,598, esto es debido a la gran diversidad de juegos que existen

**Completitud del dataset:**
_(usa `pct_celdas_ausentes()` a nivel general, y `pct_ausentes()` por columna — ¿hay alguna columna particularmente incompleta?)_

- General: 1.98% de celdas ausentes en todo el conjunto de datos (329 celdas nulas sobre un total de 16598).

**Duplicidad:**
_(usa `n_filas_duplicadas()` / `pct_filas_duplicadas()`)_

- Cero duplicadas: 0 / 0.00%

**Para cada alerta que marcaste: ¿la tratarías, la dejarías documentada sin tocar, o necesitas más contexto del origen del dato para decidir?**

- Rank: No es necesaria conservarla, ya que es un identificador solamente.
- Name: Se mantendria tal cual, ya que nos dice que juego es
